### PR TITLE
Functional: scan allowing for scalar inputs

### PR DESCRIPTION
**Pull request by nfarabullini**
Allow scalar floats in scan_operators:

```python
@scan_operator(axis=KDim, forward=True, init=(0.0))
    def _scan_scalar(carry: float, qc_in: float, scalar: float) -> float:
        qc = qc_in + carry + scalar
        return qc

@program
def scan_scalar(qc: Field[[IDim, KDim], float], scalar: float):
    _scan_scalar(qc, scalar, out=qc)
```

### CODE REVIEW

File: src/functional/iterator/embedded.py [hunk 1]
```diff
@@ -178,12 +178,13 @@ class Column(np.lib.mixins.NDArrayOperatorsMixin):
...
 
-    def __init__(self, kstart: int, data: np.ndarray) -> None:
+    def __init__(self, kstart: int, data: np.ndarray | Scalar) -> None:
         self.kstart = kstart
         self.data = data
 
     def __getitem__(self, i: int) -> Any:
-        result = self.data[i - self.kstart]
+        assert isinstance(self.data, (np.ndarray, Scalar))
+        result = self.data[i - self.kstart] if isinstance(self.data, np.ndarray) else self.data
```
Comment (havogt) [inline src/functional/iterator/embedded.py:187]: not sure in which state this PR is, but this seems wrong.

Comment (nfarabullini) [reply src/functional/iterator/embedded.py:187]: yes, I should have converted it back to draft. I discussed this change yesterday with @tehrengruber and he agreed. We  can talk about ti after the standup

File: tests/functional_tests/ffront_tests/test_execution.py [hunk 2]
```diff
@@ -1434,3 +1434,21 @@ def consume_constants(input: Field[[IDim], np.float32]) -> Field[[IDim], np.floa
     output = np_as_located_field(IDim)(np.zeros((1,), dtype=np.float32))
     consume_constants(input, out=output, offset_provider={})
     assert np.allclose(np.asarray(output), constants.PI * constants.E)
+
+
+def test_scalar_scan():
```
Comment (havogt) [inline tests/functional_tests/ffront_tests/test_execution.py:911]: please add a test with tuples of scalars, it might trigger in type_info.py:583

File: src/functional/ffront/type_info.py [hunk 3]
```diff
@@ -89,18 +89,24 @@ def function_signature_incompatibilities_fieldop(
...
-        yield "Arguments to scan operator must be fields or tuples thereof."
+    if not all(
+        is_field_type_or_tuple_of_field_type(arg) or isinstance(arg, ts.ScalarType) for arg in args
+    ):
+        yield "Arguments to scan operator must be fields, scalars or tuples thereof."
         return
 
-    arg_dims = [extract_dims(el) for arg in args for el in primitive_constituents(arg)]
+    new_args = [
+        ts.FieldType(dims=[], dtype=extract_dtype(arg)) if is_number(arg) else arg for arg in args
```
Comment (havogt) [inline src/functional/ffront/type_info.py:99]:
```suggestion
        ts.FieldType(dims=[], dtype=extract_dtype(arg)) if is_instance(arg, ts.ScalarType) else arg for arg in args
```
to be consistent with the check above?

File: src/functional/iterator/embedded.py (hunk 1 above)
Comment (havogt) [reply src/functional/iterator/embedded.py]: I'll push a slightly different implementation, making it a full array (as other functions rely on data to be an ndarray).

**Review by havogt — CHANGES_REQUESTED** [review]
I don't see if tuple of scalars will work. Can you add the test.

File: tests/functional_tests/ffront_tests/test_execution.py [hunk 4]
```diff
@@ -906,3 +907,48 @@ def consume_constants(input: Field[[IDim], np.float32]) -> Field[[IDim], np.floa
...
+
+    @field_operator
+    def scan_tuple_scalar(
+        qc: Field[[IDim, KDim], float], scalar: float
+    ) -> tuple[Field[[IDim, KDim], float], Field[[IDim, KDim], float]]:
+        return _scan_tuple_scalar(qc, scalar)
+
+    scan_tuple_scalar(qc, scalar, out=(qc, qc2), offset_provider={})
+    assert np.allclose(np.asarray(qc), expected)
+    assert np.allclose(np.asarray(qc2), -expected)
```
Comment (havogt) [inline tests/functional_tests/ffront_tests/test_execution.py:954]: I meant a tuple argument for the scan operator (since this is what you are implementing here), something like (didn't fix expected).

```suggestion
def test_tuple_scalar_scan():
    size = 10
    KDim = Dimension("K", kind=DimensionKind.VERTICAL)
    qc = np_as_located_field(IDim, KDim)(np.zeros((size, size)))
    scalar = 1.0
    expected = np.full((size, size), np.arange(start=1, stop=11, step=1).astype(float64))

    @scan_operator(axis=KDim, forward=True, init=(0.0))
    def _scan_scalar(carry: float, qc_in: float, scalar: tuple[float, float]) -> float:
        qc = qc_in + carry + scalar[0] + scalar[1]
        return qc

    @program
    def scan_scalar(qc: Field[[IDim, KDim], float], scalar: tuple[float, float]):
        _scan_scalar(qc, scalar, out=qc)

    scan_scalar(qc, (scalar, scalar), offset_provider={})
    assert np.allclose(np.asarray(qc), expected)

```

Comment (havogt) [reply tests/functional_tests/ffront_tests/test_execution.py:954]: I propose to delete the `is_field_type_or_tuple_of_field_type(type_: ts.TypeSpec)` and replace it by a more generic version
```python
def is_type_or_tuple_of_type(
    type_: ts.TypeSpec, expected_type: ts.TypeSpec | tuple[ts.TypeSpec, ...]
) -> bool:
    if not isinstance(expected_type, tuple):
        expected_type = (expected_type,)
    return all(isinstance(t, expected_type) for t in primitive_constituents(type_))
```

**Review by havogt — CHANGES_REQUESTED** [review]

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/functional/ffront/type_info.py [hunk 5]
```diff
@@ -89,18 +111,30 @@ def function_signature_incompatibilities_fieldop(
...
+    new_el: ts.TypeSpec
+    for arg_i in args:
+        if is_type_or_tuple_of_type(arg_i, ts.ScalarType):
+            if isinstance(arg_i, ts.TupleType):
+                new_el = ts.TupleType(types=unfold_scanop_tuples(arg_i, arg_i.types))
+            else:
+                new_el = ts.FieldType(dims=[], dtype=extract_dtype(arg_i))
+        else:
+            new_el = arg_i
+        new_args.append(new_el)
```
[the remark this answers is not in the document]
Comment (nfarabullini) [reply src/functional/ffront/type_info.py]: If these lines are not included, the code outputs the following error:

- Expected 1-th argument to be of type tuple[Field[[], float64], tuple[Field[[], float64], Field[[], float64]]], but got tuple[float64, tuple[float64, float64]].

../../src/functional/type_system/type_info.py:597: GTTypeError

File: src/functional/ffront/type_info.py [hunk 6]
```diff
@@ -55,17 +55,39 @@ def promote_zero_dims(
     args: list[ts.TypeSpec], function_type: ts.FieldOperatorType | ts.ProgramType
 ) -> list:
     """Promote arg types to zero dimensional fields if compatible and required by function signature."""
-    new_args = args.copy()
-    for arg_i, arg in enumerate(args):
+    current_args = args.types if isinstance(args, ts.TupleType) else args
```
[the remark this answers is not in the document]
Comment (nfarabullini) [reply src/functional/ffront/type_info.py]: thanks for spotting, it was a left-over

File: src/functional/ffront/type_info.py [hunk 7]
```diff
@@ -89,18 +103,30 @@ def function_signature_incompatibilities_fieldop(
...
+        if is_type_or_tuple_of_type(arg_i, ts.ScalarType):
+            new_el = apply_to_primitive_constituents(
+                arg_i,
+                lambda primitive_type: ts.FieldType(dims=[], dtype=extract_dtype(primitive_type)),
+            )
+        else:
+            new_el = arg_i
+        new_args.append(new_el)
+
+    arg_dims = [extract_dims(el) for arg in new_args for el in primitive_constituents(arg)]
```
Comment (tehrengruber) [inline src/functional/ffront/type_info.py:117]:
```suggestion
```
After your clarification it looks to me that we can remove this and leverage `promote_zero_dims` after we have build the `function_type`. See the suggestion below.

File: src/functional/ffront/type_info.py [hunk 8]
```diff
@@ -131,7 +157,7 @@ def _as_field(dtype: ts.ScalarType, path: tuple[int, ...]) -> ts.FieldType:
         returns=ts.DeferredType(constraint=None),
     )
 
-    yield from function_signature_incompatibilities(function_type, args, kwargs)
+    yield from function_signature_incompatibilities(function_type, new_args, kwargs)
```
Comment (tehrengruber) [inline src/functional/ffront/type_info.py:155]:
```suggestion
    yield from function_signature_incompatibilities(function_type, promote_zero_dims(args, function_type), kwargs)
```
With my current understanding this should make the changes above superfluous.

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/functional/ffront/type_info.py (hunk 7 above)
Comment (nfarabullini) [reply src/functional/ffront/type_info.py:117]: Last line is still needed but indeed the rest can be removed. Good call

File: src/functional/ffront/type_info.py [hunk 9]
```diff
@@ -52,17 +52,32 @@ def _is_zero_dim_field(field: ts.TypeSpec) -> bool:
...
-                raise GTTypeError(f"{arg} is not compatible with {def_type}.")
+        def_type = (
+            function_type.args[arg_i]
+            if isinstance(function_type, ts.FunctionType)
+            else function_type.definition.args[arg_i]
+        )
+
+        def _as_field(arg: ts.TypeSpec, path: tuple):
+            try:
+                def_t = reduce(lambda type_, idx: type_.types[idx], path, def_type)  # noqa: B023
```
Comment (tehrengruber) [inline src/functional/ffront/type_info.py:68]:
```suggestion
                el_def_type = reduce(lambda type_, idx: type_.types[idx], path, def_type)  # noqa: B023
```
How about `el_def_type`? Or alternatively rename the outer `def_type` to `outer_def_type`. Just `def_t` however is somewhat against our variable naming conventions.

**Review by tehrengruber — APPROVED** [review]

File: src/functional/ffront/type_info.py [hunk 10]
```diff
@@ -52,17 +52,34 @@ def _is_zero_dim_field(field: ts.TypeSpec) -> bool:
 
 
 def promote_zero_dims(
-    args: list[ts.TypeSpec], function_type: ts.FieldOperatorType | ts.ProgramType
+    args: list[ts.TypeSpec], function_type: ts.FieldOperatorType | ts.ProgramType | ts.FunctionType
 ) -> list:
     """Promote arg types to zero dimensional fields if compatible and required by function signature."""
-    new_args = args.copy()
+    new_args = args
```
Comment (havogt) [inline src/functional/ffront/type_info.py:58]: Why this change? Aren't we modifying the original args now?

File: src/functional/ffront/type_info.py [hunk 11]
```diff
@@ -52,17 +52,34 @@ def _is_zero_dim_field(field: ts.TypeSpec) -> bool:
...
+            else function_type.definition.args[arg_i]
+        )
+
+        def _as_field(arg: ts.TypeSpec, path: tuple):
+            try:
+                el_def_type = reduce(
+                    lambda type_, idx: type_.types[idx], path, def_type  # noqa: B023
+                )
+            except (IndexError, AssertionError):
+                return el_def_type
```
Comment (havogt) [inline src/functional/ffront/type_info.py:72]: Can we give this a name? I have zero intuition for what it does?

File: src/functional/iterator/embedded.py [hunk 12]
```diff
@@ -702,7 +707,7 @@ def _get_axes(
 
 
 def _make_tuple(
-    field_or_tuple: LocatedField | tuple,  # arbitrary nesting of tuples of LocatedField
+    field_or_tuple: LocatedField | tuple | ConstantField,
```
Comment (havogt) [inline src/functional/iterator/embedded.py:710]: This change should not be necessary. ConstantField should be a LocatedField.

File: src/functional/type_system/type_info.py [hunk 13]
```diff
@@ -227,22 +227,22 @@ def is_arithmetic(symbol_type: ts.TypeSpec) -> bool:
...
-     >>> is_field_type_or_tuple_of_field_type(ts.TupleType(types=[field_type, field_type]))
-     True
-     >>> is_field_type_or_tuple_of_field_type(ts.TupleType(types=[field_type, scalar_type]))
-     False
+    >>> is_type_or_tuple_of_type(field_type, ts.FieldType)
+    True
+    >>> is_type_or_tuple_of_type(ts.TupleType(types=[scalar_type, field_type]), (ts.ScalarType, ts.FieldType))
+    True
+    >>> is_type_or_tuple_of_type(scalar_type, ts.FieldType)
+    False
```
Comment (havogt) [inline src/functional/type_system/type_info.py:243]:
```suggestion
    False
    >>> is_type_or_tuple_of_type(ts.TupleType(types=[scalar_type, field_type]), ts.FieldType)
    False
```
just for completness

**Review by havogt — CHANGES_REQUESTED** [review]
The promote_zero_dims, requires some clarification. Rest looks good.

**Review by havogt — APPROVED** [review]
Thanks! I think it looks good now.
